# Problems Notebook

This notebook contains my work of the assessment in Applied Statistics module in the Higher Diploma in Data Analytics course. 
The notebook covers four main problems: 




## Problem 1: Extending the Lady Tasting Tea:

Let's extend the Lady Tasting Tea experiment as follows. The original experiment has 8 cups: 4 tea-first and 4 milk-first. Suppose we prepare 12 cups: 8 tea-first and 4 milk-first. A participant claims they can tell which was poured first.

Simulate this experiment using numpy by randomly shuffling the cups many times and calculating the probability of the participant correctly identifying all cups by chance. Compare your result with the original 8-cup experiment.

In your notebook, explain your simulation process clearly, report and interpret the estimated probability, and discuss whether, based on this probability, you would consider extending or relaxing the p-value threshold compared to the original design.

### Solution to the Problem 1

In [42]:
# Import necessary libraries

# for mathematical functions
# see https://docs.python.org/3/library/math.html
import math

# for combinatorial functions
# see https://docs.python.org/3/library/itertools.html
import itertools

# for random number generation
# see https://www.w3schools.com/python/module_random.asp
import random

# for numerical operations
# see https://www.w3schools.com/python/numpy/numpy_intro.asp
import numpy as np

# for data visualization and plotting
# see https://www.w3schools.com/python/matplotlib_pyplot.asp
import matplotlib.pyplot as plt

#### Mathematical Concepts

https://docs.python.org/3/library/math.html#math.comb

#### Cups of Tea in Original Experiment

In [43]:
# Number of cups of tea. All cups are identical. 
num_cups= 8

# Number of cups with tea first
num_tea_first = 4

# Number of cups with milk first
num_milk_first= 4

In [44]:
# Number of ways to select 8 cups from 12

tea_ways = math.comb(num_cups, num_tea_first)

print(f'Number of ways to select cups is {tea_ways}')

Number of ways to select cups is 70


#### Cups of Tea in Extended Experiment

In [45]:
# Number of cups of tea. All cups are identical. 
num_cups_ex = 12

# Number of cups with tea first
num_tea_first_ex = 8

# Number of cups with milk first
num_milk_first_ex = 4

In [46]:
# Number of ways to select 8 cups from 12

tea_ways_ex = math.comb(num_cups_ex, num_tea_first_ex)

print(f'Number of ways to select cups is {tea_ways_ex}')

Number of ways to select cups is 495


In [47]:
# number of ways to select 4 cups from 12
milk_ways_ex = math.comb(num_cups_ex, num_milk_first_ex)

print(f'Number of ways to select cups is {milk_ways_ex}')

Number of ways to select cups is 495


Both results shows 495, however these results are not exaclty the same. 

If we choose 8 cups to have tea first, it leaves us with remaining 4 cups with milk. If we choose 4 cups with milk, we will have 8 cups remaining. 

Let's check that carefully. 

##### Select 8 cups to have tea first.

Label the 12 cups of tea from the extened experiment `1`, `2`, `3`, `4`, `5`, `6`, `7`, `8`, `9`, `10`, `11`, `12`. 

We select 8 cups to have tea in first.

For example, we might select `1`, `2`, `4`, `6`, `9`, `10`, `11` and `12`.

Visualize having eight saucers where we place eight of the cups.

How many ways are there to select 8 cups from 12?

Cups: `1 2 3 4 5 6 7 8 9 10 11 12`

Saucers: `_ _ _ _ _ _ _ _`

In [48]:
# Total number of ways of selecting 8 cups from 12, keeping them in order.
ways_ordered_tea = 12 * 11 * 10 * 9 * 8 * 7 * 6 * 5

# Show.
ways_ordered_tea

19958400

The order doesn't matter. If cups `1`, `2`, `4`, `6`, `9`, `10`, `11` and `12`. are selected in order it is the same result as if `12`, `11`, `9`, `10`, `2`, `6`, `4` and `1`. were selected, or any other permutation.

`1 2 4 6 9 10 11 12`

`12 11 9 10 2 6 4`

`_ _ _ _ _ _ _ _`

In [49]:
# No of ways of shuffling 8 cups.
no_shuffles_tea = 8 * 7 * 6 * 5 * 4 * 3 * 2 * 1

# Show.
no_shuffles_tea

40320

In [50]:
# No of combinations ups with tea.
no_comb_tea = ways_ordered_tea // no_shuffles_tea

# Show.
no_comb_tea

495

##### Select 4 cups to have milk first.

Again, we have our labeled 12 cups of tea `1`, `2`, `3`, `4`, `5`, `6`, `7`, `8`, `9`, `10`, `11`, `12`. 

But now we select 4 cups to have milk in first. We start again, so we still have 12 cups.

Let's select `3`, `5`, `7`, and `8`.

We place them on the other four saucers where we place four cups on, where we guess the milk is first. 

How many ways are there to select 4 cups from 12?

Cups: `1 2 3 4`

Saucers: `_ _ _ _`

In [51]:
# Total number of ways of selecting 4 cups from 12, keeping them in order.
ways_ordered_milk = 12 * 11 * 10 * 9

# Show.
ways_ordered_milk

11880

And again, the order doesn't matter. 

In [52]:
# No of ways of shuffling four cups.
no_shuffles_milk = 4 * 3 * 2 * 1

# Show.
no_shuffles_milk

24

In [53]:
# No of combinations milk.
no_combs = ways_ordered_milk // no_shuffles_milk

# Show.
no_combs

495

There are 495 ways to select 8 cups with tea, which is similar as choosing 4 cups with milk. It is just viewed from the opposite perspective.


In mathematics, the `math.comb()` function is called the *choose* function.

It is written and calculated as

$\binom{n}{k} = \frac{n!}{k!(n-k)!}$.

In [54]:
# Number of cups of tea in total.
n = 12

In [55]:
# Number of cups of tea with milk in first.
k = 8

## math.factorial

https://docs.python.org/3.12/library/math.html#math.factorial

Mathematically denoted as `n!`.

Multiply all positive integers from `n` down to 1.


`n! = n x (n - 1) x (n - 2) x ... x 1`

e.g. `5! = 5 x 4 x 3 x 2 x 1 = 120`

In [56]:
# Twelve factorial.
math.factorial(n)

479001600

In [58]:
# Four factorial.
math.factorial(k)

40320

In [59]:
# Four factorial.
math.factorial(n - k)

24

In [60]:
# No of ways of selecting k objects from n without replacement and without order.
math.factorial(n) // (math.factorial(k) * math.factorial(n - k))

495

## Problem 2: Normal Distribution

Generate 100,000 samples of size 10 from the standard normal distribution. For each sample, compute the standard deviation with ddof=1 (sample SD) and with ddof=0 (population SD). Plot histograms of both sets of values on the same axes with transparency. Describe the differences you see. Explain how you expect these differences to change if the sample size is increased.

## Problem 3: t-Tests

A type II error occurs when a test fails to reject the null hypothesis even though it is false. For each mean difference
$ d =0,0.1,0.2,…,1.0 $, repeat the following simulation 1,000 times:


* Draw two samples of size 100, one from the standard normal distribution and one from the normal distribution with mean $d$ and standard deviation 1.
* Run an independent samples t-test on the two samples, rejecting the null hypothesis if the p-value is less than 0.05.
* Record the proportion of times the null hypothesis is not rejected.


Plot this proportion against $d$, and explain how the type II error rate changes as the difference in means increases.

## Problem 4: ANOVA

Generate three independent samples, each of size 30, from normal distributions with means 0, 0.5, and 1, each with standard deviation 1.

* Perform a one-way ANOVA to test whether all three means are equal.
* Perform three independent two-sample t-tests: samples 1 vs 2, 1 vs 3, and 2 vs 3.
* Compare the conclusions.

Write a short note on why ANOVA is preferred over running several t-tests.

Importing nessesary libraries

In [57]:
import numpy as np

***
End